### PR TITLE
Passing report class object to celery task

### PR DESCRIPTION
**Pull request by Rohit25negi**
1. instead of sending request & config now sending report class object

http://manage.dimagi.com/default.asp?281230#1518301

### CODE REVIEW

**Comment by Rohit25negi** [comment]
@orangejenny @calellowitz  this one is urgent. please have a look when you have time :)

**Review by calellowitz — APPROVED** [review]
Good catch. Thanks for the quick fix

**Review by emord — COMMENTED** [review]
looks fine for now since it's urgent. I think in general it's bad practice to send full objects to celery as noted here: https://docs.google.com/presentation/d/1iiiVZDiOGXoLeTvEIgM_rGgw6Me5_wM_Cyc64bl7zns/edit#slide=id.g1d621cb6fc_0_151

I don't have enough context on how large `ReportConfig` is, but I think you could pass in its id instead and the celery task could use `ReportConfig.get(report_config_id)` or something similar

**Comment by Rohit25negi** [comment]
@esoergel oww i see. Now I got the reason for passing the report state and class in the celery task and again make the object. initialy I thought it is just extra step. Now I understood. Thanks for pointing that out.
PS: Update the code to send the report state instead of entire object.

**Comment by esoergel** [comment]
Hey @Rohit25negi , there are a couple issues with this approach - I understand these were introduced in the previous PR, but I think they warrant addressing.   This generates a report, attempts to send it, and if there's an error, it then generates the report from scratch again.  This means that the biggest reports are built twice.  Also, As-is, this passes the email body through as a parameter to a celery task - @emord mentioned the issue with passing large objects to celery, and the report body is very very large, which I imagine could overwhelm RabbitMQ.

The simplest solution would be to always send a download link (though it's up to product whether that's acceptable), then there'd be no need for `send_email_report`.

If that's not an option, I'd make a task similar to `export_all_rows_task` which handles generating the report, attempting to send it, and failing that, sends saves it in redis and sends a download link.  This would push the report generation off the main thread, and avoid passing around large objects or generating them twice.

**Comment by calellowitz** [comment]
@esoergel I am not sure that approach is going to work. Specifically this PR (and the previous one) deals with failures of the reports that are rendered as the body of an email in html. I think after that failure we will always need to rerender as a csv/xls instead of the failed html, and as far as I can tell that is only possible by rerendering the report without significant refactoring of our report code. This is a super weird edge case (as far as I know the only reports that this even happens are from enikshay who tried to send massive reports in the body of the email, most people just use the export functionality at scale), and while it is not efficient to build the report twice, I am not sure the difference between 5 and 10 minutes (or more likely 2 and 4 minutes) matters in this case, since there is an expectation of waiting. 

As far as passing the body itself, that is something we do for all async emails, and were doing for report emails before this PR.  If it is an easy fix, I think it is worth changing to render the html body in `send_email_report`, but if not, I do not think it should block this PR.

I am against always sending a download link, at least without product signoff and potentially a large look at use cases, since the entire purpose of the email feature is that it is rendered in the email. If people wanted an export, they would use the export button instead. That solution just removes this feature entirely.

**Comment by Rohit25negi** [comment]
Was looking into the code, It doesn't seem to be an easy fix for  this PR as it will need deeper changes in the code base to implement them in a good way.

**Comment by esoergel** [comment]
@Rohit25negi I agree that that doesn't need to block this PR.

@calellowitz do you know if the `email_queue` queue is somehow exempt from the issue with storing large parameters?  At the very least I think this task should go in that queue.  I don't think it would be challenging to move the report generation into an async task that also sends the email.  This would avoid passing anything large and move the process off the main thread.

**Comment by Rohit25negi** [comment]
@calellowitz @esoergel  I have moved the report generation into the async task.  It also prevents passing big email bodies to `send_email_report` function.

File: corehq/apps/reports/tasks.py
Comment (esoergel) [inline corehq/apps/reports/tasks.py]: oof - this is a lot more involved than I was expecting - sorry for sending you down this path.

**Review by esoergel — APPROVED** [review]
thanks for the followup

File: corehq/apps/reports/tasks.py
Comment (Rohit25negi) [reply corehq/apps/reports/tasks.py]: yes It involved a little more. But its cool, I got to learn more bits of code. 
#CheersToThat 👍

**Comment by Rohit25negi** [comment]
hey, If changes looks cool. shall I merge this PR?

**Comment by calellowitz** [comment]
Yep. Once you get the green check mark for approved from whoever gave feedback, feel free to merge

**Pull request merged**
